### PR TITLE
Feature/jagtrrz/bulk certificates

### PR DESCRIPTION
**Pull request by jagtrrz**

### CODE REVIEW

File: breathecode/certificate/tasks.py [hunk 1]
```diff
@@ -54,3 +54,25 @@ def generate_cohort_certificates(self, cohort_id):
...
+
+@shared_task(bind=True, base=BaseTaskWithRetry)
+def generate_one_certificate(self, cohort_id, user_id):
+    logger.debug("Starting generate_cohort_certificates")
+    from .actions import generate_certificate
+
+    cohort_user =  CohortUser.objects.filter(cohort__id=cohort_id, 
+            user_id=user_id, role='STUDENT').first()
+
+    print("@@@@@@@@@@@@@", cohort_user)
```
Comment (jefer94) [inline breathecode/certificate/tasks.py:66]: Before each commit, we need remove all the prints for the code

File: breathecode/certificate/tasks.py [hunk 2]
```diff
@@ -54,3 +54,25 @@ def generate_cohort_certificates(self, cohort_id):
...
+    if not cohort_user:
+        logger.error(f'Cant generate certificate with {user_id}')
+        return
+    
+
+    logger.debug(f"Generating gertificate for {str(cohort_user)} students that GRADUATED")
```
Comment (jefer94) [inline breathecode/certificate/tasks.py:72]: Check if it print show the `cohort_user` good

File: breathecode/certificate/tests/urls/tests_root.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        response = self.client.post(url, data, format='json')
+        json = response.json()
+        expected = {
+            'detail': "You (user: 1) don't have this capability: crud_certificate for academy 1",
+            'status_code': 403
+        }
+
+        self.assertEqual(json, expected)
+        self.assertEqual(response.status_code, 403)
+        self.assertEqual(self.all_user_specialty_dict(), [])
```
Comment (jefer94) [inline breathecode/certificate/tests/urls/tests_root.py:112]: We need any test that generate any certificate, for now it not exist

File: breathecode/certificate/views.py [hunk 4]
```diff
@@ -113,6 +115,7 @@ def post(self, request, cohort_id, academy_id=None):
         return Response(all_certs, status=status.HTTP_201_CREATED)
 
 
+
```
[the remark this answers is not in the document]
Comment (jefer94) [reply breathecode/certificate/views.py:118]: Use one or two return character instead of three

File: breathecode/certificate/views.py [hunk 5]
```diff
@@ -134,7 +137,33 @@ def get(self, request, academy_id=None):
             return self.get_paginated_response(serializer.data)
         else:
             return Response(serializer.data, status=status.HTTP_200_OK)
+    
+    @capable_of('crud_certificate')
+    def post(self, request, academy_id=None):
+        is_many = isinstance(request.data, list)
+        data = request.data
```
Comment (jefer94) [inline breathecode/certificate/views.py:144]: If i get just one element, transform data to one list of one element

File: breathecode/certificate/views.py [hunk 6]
```diff
@@ -134,7 +137,33 @@ def get(self, request, academy_id=None):
...
+        cohort_users = []
+      
+        if is_many:
+            for items in data:
+                cohort__slug = items.get("cohort_slug")
+                user__id = items.get("user_id")
+                cohort_user =  CohortUser.objects.filter(cohort__slug=cohort__slug, 
+                        user_id=user__id, role='STUDENT', cohort__academy__id=academy_id).first()
+                if cohort_user is not None :
```
Comment (jefer94) [inline breathecode/certificate/views.py:153]: One extra space character

**Review by jefer94 — CHANGES_REQUESTED** [review]

File: breathecode/certificate/tests/urls/tests_root.py (hunk 3 above)
Comment (jagtrrz) [reply breathecode/certificate/tests/urls/tests_root.py:112]: Este test lo creo que la issue https://github.com/4GeeksAcademy/About-4Geeks-Academy/issues/1284 que es donde he cambiado las condiciones del POST.

Comment (jefer94) [reply breathecode/certificate/tests/urls/tests_root.py:112]: We need the test here in this pull request, try to provide de basic data to check that the endpoint work, check it checking you can pass a valid request and in the db it make a certificate with the correct status

**Review by jefer94 — CHANGES_REQUESTED** [review]

**Comment by jefer94** [comment]
We need the test here in this pull request, try to provide de basic data to check that the endpoint work, check it checking you can pass a valid request and in the db it make a certificate with the correct status

File: breathecode/certificate/views.py [hunk 7]
```diff
@@ -134,7 +135,37 @@ def get(self, request, academy_id=None):
...
+        cohort_users = []
+
+        if len(data) > 0:
+            for items in data:
+                cohort__slug = items.get("cohort_slug")
+                user__id = items.get("user_id")
+                cohort_user =  CohortUser.objects.filter(cohort__slug=cohort__slug, 
+                        user_id=user__id, role='STUDENT', cohort__academy__id=academy_id).first()
+
+                if cohort_user is not None :
```
Comment (jefer94) [inline breathecode/certificate/views.py:155]: Change `if cohort_user is not None :` to `if cohort_user is not None:`

File: breathecode/certificate/tests/actions/tests_generate_certificate.py [hunk 8]
```diff
@@ -10,6 +10,7 @@
 from ...actions import generate_certificate, strings
 from ..mixins.new_certificate_test_case import CertificateTestCase
 from ....admissions.models import FULLY_PAID, UP_TO_DATE, LATE
+from ...tasks import generate_one_certificate
```
Comment (jefer94) [inline breathecode/certificate/tests/actions/tests_generate_certificate.py:13]: You can remove this unused import

File: breathecode/certificate/tests/urls/tests_root.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            'user_id':  model['user'].id,
+        }]
+        response = self.client.post(url, data, format='json')
+        json = response.json()
+        expected = {
+            'detail': "You (user: 1) don't have this capability: crud_certificate for academy 1",
+            'status_code': 403
+        }
+
+        self.assertEqual(json, expected)
```
Comment (jefer94) [inline breathecode/certificate/tests/urls/tests_root.py:110]: We need a test with a successful response, this test suite just have 4xx cases

**Review by jefer94 — CHANGES_REQUESTED** [review]

**Comment by jefer94** [comment]
check it https://github.com/breatheco-de/apiv2/pull/69/checks?check_run_id=2797847726

File: breathecode/certificate/models.py [hunk 10]
```diff
@@ -108,7 +109,7 @@ def clean(self):
         self.is_cleaned = True
 
     def save(self, *args, **kwargs):
-
+        print("hello world")
```
Comment (jefer94) [inline breathecode/certificate/models.py:112]: remove all prints

File: breathecode/certificate/urls.py [hunk 11]
```diff
@@ -13,6 +13,6 @@
     path('token/<str:token>/', get_certificate),
     
     path('cohort/<int:cohort_id>/student/<int:student_id>', CertificateView.as_view()),
-    path('cohort/<int:cohort_id>', CertificateCohortView.as_view(), name="certificate_cohort"),
-    path('', CertificateAcademyView.as_view()),
+    path('cohort/<int:cohort_id>', CertificateCohortView.as_view(),  name="certificate_cohort"),
```
Comment (jefer94) [inline breathecode/certificate/urls.py:16]: too many spaces

**Comment by jefer94** [comment]
Remove this file https://github.com/breatheco-de/apiv2/blob/4751a148f5de22095670220ffdbf11183c8baed8/temp-20988.rdb

**Review by jefer94 — CHANGES_REQUESTED** [review]

File: breathecode/certificate/tests/urls/tests_root.py [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+        }
+        self.assertEqual(json, expected)
+        self.assertEqual(response.status_code, 404)
+        self.assertEqual(self.all_user_specialty_dict(), [])
+
+    
+    @patch(GOOGLE_CLOUD_PATH['client'], apply_google_cloud_client_mock())
+    @patch(GOOGLE_CLOUD_PATH['bucket'], apply_google_cloud_bucket_mock())
+    @patch(GOOGLE_CLOUD_PATH['blob'], apply_google_cloud_blob_mock())
+    def test_certificate_re_attemps(self):
```
Comment (jefer94) [inline breathecode/certificate/tests/urls/tests_root.py:143]: We need a new test like it that generate two UserSpecialty, to certificate the view accept bulk mode

**Review by jefer94 — CHANGES_REQUESTED** [review]

**Review by jefer94 — APPROVED** [review]

**Comment by jefer94** [comment]
@alesanchezr is done

**Pull request merged**